### PR TITLE
[MODEL-7883] updates based on cfds feedback

### PR DESCRIPTION
**Pull request by drenfr01**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary

- Removed repeated docstrings on everything but the "simple" task templates that are designed as tutorials, i.e. we kept the docstrings for fit() and predict() for a basic sklearn binary estimator but removed it for the keras regression template
- Added in a jupyter notebook tutorial to demonstrate how to run a custom task
- updated some minor error messages in model_adapter
- provided more extensive docstring comments on how we handle the class labels for binary / multiclass estimators
- ensured all new templates have model metadata file
- minor other fixes

## Rationale

This is primarily driven by a feedback session with our CFDS stakeholder, Eric Bolme. He acted as an internal persona for the "code first data scientist" and provided insights on easy it was to run our drum in a notebook, what pieces of code required additional comments, etc.

### CODE REVIEW

File: custom_model_runner/datarobot_drum/drum/model_adapter.py [hunk 1]
```diff
@@ -685,9 +685,12 @@ def fit(self, X, y, output_dir, class_order=None, row_weights=None, parameters=N
                         parameters=parameters,
                     )
 
-                    self._custom_task_class_instance.save(self._model_dir)
+                    try:
+                        self._custom_task_class_instance.save(self._model_dir)
+                    except AttributeError:
+                        raise DrumCommonException("There appears to be an issue with your save hook in custom.py")
```
Comment (drenfr01) [inline custom_model_runner/datarobot_drum/drum/model_adapter.py:691]: change this to except Exception

File: tests/fixtures/custom_task_interface_keras_visual/model_utils.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+    # extract preprocessor pipeline
+    preprocessor = estimator_pipeline[:-1]
+
+    # extract keras model from the pipeline obj
+    keras_model = estimator_pipeline[-1]
+
+    # save the model (in '.h5' format - as its easy to load using keras' load_model() later)
+    # to BytesIO obj (in RAM - as its easy to joblib dump later)
+    io_container = io.BytesIO()
+    with h5py.File(io_container, mode="w") as file:
```
Comment (jmbannon) [inline tests/fixtures/custom_task_interface_keras_visual/model_utils.py:293]: Use `save_model`, `load_model` function

**Comment by jmbannon** [comment]
please add a description of what's changed

File: custom_model_runner/datarobot_drum/custom_task_interfaces/custom_task_interface.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 
 class Serializable(object):
-    default_filename = "artifact.pkl"
+    default_artifact_filename = "artifact.pkl"
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/custom_task_interfaces/custom_task_interface.py:6]: We should change the filename to something like `drum_artifact.pkl`, something that users probably wont overwrite

File: custom_model_runner/datarobot_drum/drum/model_adapter.py [hunk 4]
```diff
@@ -685,9 +685,16 @@ def fit(self, X, y, output_dir, class_order=None, row_weights=None, parameters=N
                         parameters=parameters,
                     )
 
-                    self._custom_task_class_instance.save(self._model_dir)
+                    try:
+                        self._custom_task_class_instance.save(self._model_dir)
+                    except Exception:
+                        raise DrumCommonException(
+                            "There appears to be an issue with your save hook in custom.py"
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/model_adapter.py:692]: Would change this to be `An error occurred when saving your custom task. Ensure all variables stored on self are pickleable or excluded when saving`

File: custom_model_runner/datarobot_drum/drum/model_adapter.py [hunk 5]
```diff
@@ -685,9 +685,16 @@ def fit(self, X, y, output_dir, class_order=None, row_weights=None, parameters=N
...
+                        )
                 except AttributeError:
-                    raise DrumCommonException("Fit method must be defined in your custom.py file")
+                    raise DrumCommonException(
+                        "There appears to be an issue with your fit hook in custom.py"
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/model_adapter.py:696]: We should leave the original message intact

File: tests/fixtures/custom_task_interface_anomaly/custom.py [hunk 6]
```diff
@@ -14,48 +14,13 @@
...
-        X: pd.DataFrame
-            Data that DataRobot passes for transformation.
-
-        Returns
-        -------
-        pd.DataFrame
-            Returns a dataframe with transformed data.
-        """
 
+        # Note how anomaly estimators only use one column, so no explicit column names are needed
```
Comment (jmbannon) [inline tests/fixtures/custom_task_interface_anomaly/custom.py:25]: `only output one column`

File: tests/fixtures/custom_task_interface_binary/custom.py [hunk 7]
```diff
@@ -57,4 +59,5 @@ def predict_proba(self, X, **kwargs):
             Returns a dataframe with transformed data.
         """
 
+        # Note that binary estimators require 2 columns, one for each class.
```
Comment (jmbannon) [inline tests/fixtures/custom_task_interface_binary/custom.py:62]: Comment about column names

File: tests/fixtures/custom_task_interface_keras_regression/custom.py [hunk 8]
```diff
@@ -81,29 +78,12 @@ def load(cls, artifact_directory):
         cls
             The deserialized object
         """
-
         with open(Path(artifact_directory) / "artifact.pkl", "rb") as fp:
             custom_task = pickle.load(fp)
 
-        custom_task.estimator.model = keras.models.load_model(Path(artifact_directory) / "model")
```
Comment (jmbannon) [inline tests/fixtures/custom_task_interface_keras_regression/custom.py:88]: this failed before because you saved using `model.save` (TF way of doing it), then loading with `keras.models.load_model` (old keras way of doing it), fyi

File: tests/fixtures/custom_task_interface_multiclass/custom.py [hunk 9]
```diff
@@ -53,4 +55,6 @@ def predict_proba(self, X, **kwargs):
         pd.DataFrame
             Returns a dataframe with transformed data.
         """
+
+        # Note how multiclass estimators require one column per class
```
Comment (jmbannon) [inline tests/fixtures/custom_task_interface_multiclass/custom.py:59]: add comment about column names

File: tests/fixtures/custom_task_interface_transform_missing_values/custom.py [hunk 10]
```diff
@@ -11,9 +11,50 @@
...
+            Project's target column (None is passed for unsupervised projects). This allows users to implement
+            transformers that rely on the target values, e.g. target encoding
+
+        Returns
+        -------
+        CustomTask
+            returns an object instance of class CustomTask that can be used in chained method calls
+        """
+
+        # Note that for transforms we typically save everything used in the transform hook below to self
```
Comment (jmbannon) [inline tests/fixtures/custom_task_interface_transform_missing_values/custom.py:38]: This comment sounds confusing. It should be something like `Any information derived from the training data (ie median values for each column) should be stored to self. Then, in the transform hook below, we use this information to transform any data that passes through this task.`

File: custom_model_runner/datarobot_drum/drum/model_adapter.py (hunk 5 above)
Comment (drenfr01) [reply custom_model_runner/datarobot_drum/drum/model_adapter.py:696]: @jmbannon why? It's really misleading, because I would get that message often when a fit hook was actually defined and just had an error

Comment (jmbannon) [reply custom_model_runner/datarobot_drum/drum/model_adapter.py:696]: Hmmm, we should rethink this catch then. Maybe not even catch anything, so the stacktrace is intact

File: tests/fixtures/custom_task_interface_multiclass/custom.py (hunk 9 above)
Comment (drenfr01) [reply tests/fixtures/custom_task_interface_multiclass/custom.py:59]: @jmbannon sorry, which comment?

Comment (jmbannon) [reply tests/fixtures/custom_task_interface_multiclass/custom.py:59]: i mean on this line mention something about how we have to pass column names derived from the estimator

File: custom_model_runner/datarobot_drum/drum/model_adapter.py (hunk 5 above)
Comment (drenfr01) [reply custom_model_runner/datarobot_drum/drum/model_adapter.py:696]: Hmm, I don't know if I love that either. I'm going to combine the two comments as an interim step, i.e. something like "There appears to be an issue with your fit method OR it is missing entirely. Please verify that that the fit method in custom.py is present and runs without errors."

Comment (drenfr01) [reply custom_model_runner/datarobot_drum/drum/model_adapter.py:696]: Then I'm going to create a ticket to look into better logging and add this catch method in it, because we've talked several times about enhancing the logging anyway

**Review by jmbannon — APPROVED** [review]

**Review by scottp-ml — APPROVED** [review]
